### PR TITLE
Feature/roe 1653 m store trust data in session tell us about the individual page

### PR DESCRIPTION
**Pull request by okoye-companieshouse**
### JIRA link

https://companieshouse.atlassian.net/browse/ROE-1653

### Change description

Store trust data in session tell us about the individual page

### Work checklist

- [X] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
feat:  Store trust data in session tell us about the individual page
test: Test mapping service for trustees
```

### CODE REVIEW

File: src/controllers/beneficial.owner.type.controller.ts [hunk 1]
```diff
@@ -1,16 +1,13 @@
 import { NextFunction, Request, Response } from "express";
```
Comment (JoshFen95) [inline src/controllers/beneficial.owner.type.controller.ts:1]: `src/controllers/beneficial.owner.type.controller.ts` is unrelated to this story

File: src/controllers/trust.legal.entity.beneficial.owner.controller.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ type TrustIndividualBeneificalOwnerPageProperties = {
   templateName: string;
   pageData: {
     trustData: CommonTrustData,
-    individualTrusteeType: typeof RoleWithinTrustType;
+    RoleWithinTrustType: typeof RoleWithinTrustType;
```
Comment (JoshFen95) [inline src/controllers/trust.legal.entity.beneficial.owner.controller.ts:19]: I'm currently making changes to this file in my story. THis being one of them 

` roleWithinTrustType: typeof RoleWithinTrustType;`

File: src/model/beneficial.owner.type.model.ts [hunk 3]
```diff
@@ -1,12 +0,0 @@
-export const BeneficialOwnerTypeKey = "beneficial_owner_type";
-
-export enum BeneficialOwnerTypeChoice {
-    individual = "individualOwner",
-    otherLegal = "otherLegalOwner",
```
Comment (JoshFen95) [inline src/model/beneficial.owner.type.model.ts:5]: `src/model/beneficial.owner.type.model.ts` is unrelated to the current trust work. I don't think the file should be removed/split at this point as it's out of scope for the current task

File: src/model/trust.model.ts [hunk 4]
```diff
@@ -9,15 +10,23 @@ export const TrustKey = "trusts";
...
+  'trust_id',
+  'trust_name',
+  'creation_date_day',
+  'creation_date_month',
+  'creation_date_year',
+  'unable_to_obtain_all_trust_info',
+  'INDIVIDUALS',
+  'HISTORICAL_BO',
+  'CORPORATES',
 ];
```
Comment (JoshFen95) [inline src/model/trust.model.ts:22]: What does this change do?

File: src/model/trust.model.ts [hunk 5]
```diff
@@ -31,10 +40,11 @@ export interface Trust {
   INDIVIDUALS?: TrustIndividual[];
   HISTORICAL_BO?: TrustHistoricalBeneficialOwner[];
   CORPORATES?: TrustCorporate[];
+  TRUSTEES?: (GeneralTrustee | InterestedPersonTrustee)[];
```
Comment (JoshFen95) [inline src/model/trust.model.ts:43]: This is unrequired. The individual trustees from the trust individual beneficial owner page are stored in `INDIVIDUALS?: TrustIndividual[];`.

File: src/model/trust.model.ts [hunk 6]
```diff
@@ -78,6 +88,41 @@ interface TrustHistoricalBeneficialOwnerCommon {
...
+  country: string;
+  postal_code: string;
+  correspondence_address_line1?: string;
+  correspondence_property_name?: string;
+  correspondence_address_line2?: string;
+  correspondence_city?: string;
+  correspondence_county?: string;
+  correspondence_country?: string;
+  correspondence_postal_code?: string;
+}
```
Comment (JoshFen95) [inline src/model/trust.model.ts:115]: I don't think we need this as we need to map to the `TrustIndividual`  which  exists in this file. I think it needs to be a `type` rather than an `interface` though.

File: src/utils/enum/beneficial.owner.type.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export enum BeneficialOwnerType {
+    Individual = "Individual Owner",
+    OtherLegal = "Other Legal Owner",
```
Comment (JoshFen95) [inline src/utils/enum/beneficial.owner.type.ts:3]: I feel likes it's out of scope

File: src/utils/enum/managing.officer.type.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export enum ManagingOfficerType {
+    Individual = "Individual Officer",
```
Comment (JoshFen95) [inline src/utils/enum/managing.officer.type.ts:2]: I feel likes it's out of scope

File: src/utils/trust/beneficial.owner.mapper.ts [hunk 9]
```diff
@@ -1,15 +1,15 @@
...
+import { BeneficialOwnerType } from '../enum/beneficial.owner.type';
 
 const mapBoIndividualToPage = (
   beneficialOwner: BeneficialOwnerIndividual,
 ): Page.TrustBeneficialOwnerListItem => {
   return {
     id: beneficialOwner.id,
     name: [beneficialOwner.first_name, beneficialOwner.last_name].join(' '),
-    type: BeneficialOwnerTypeChoice.individual,
+    type: BeneficialOwnerType.Individual,
```
Comment (JoshFen95) [inline src/utils/trust/beneficial.owner.mapper.ts:12]: I feel likes it's out of scope

File: src/utils/trust/beneficial.owner.mapper.ts [hunk 10]
```diff
@@ -19,7 +19,7 @@ const mapBoOtherToPage = (
   return {
     id: beneficialOwner.id,
     name: beneficialOwner.name!,
-    type: BeneficialOwnerTypeChoice.otherLegal,
+    type: BeneficialOwnerType.OtherLegal,
```
Comment (JoshFen95) [inline src/utils/trust/beneficial.owner.mapper.ts:22]: I feel likes it's out of scope

File: src/utils/trust/individual.trustee.mapper.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    country: formData.country,
+    postal_code: formData.postal_code,
+    correspondence_property_name: formData.correspondence_property_name,
+    correspondence_address_line1: formData.correspondence_address_line1,
+    correspondence_address_line2: formData.correspondence_address_line2,
+    correspondence_city: formData.correspondence_city,
+    correspondence_county: formData.correspondence_county,
+    correspondence_country: formData.correspondence_country,
+    correspondence_postal_code: formData.correspondence_postal_code,
+  };
```
Comment (JoshFen95) [inline src/utils/trust/individual.trustee.mapper.ts:33]: The mapped object needs to map to `TrustIndividual` in `Trust.model.ts`. Some fields may need to be added such as second_nationality but others should be mapped directly to existing values where possible. 
Some of the fields are not obvious to what they are(and could do with altering) but for now using the excel spreadsheet and creating a json object with it can help. I can show you what I did.

We need to map to the existing object where possible to minimise effects down stream in the overseas-api and nodesdk for example. If we introduce new fields, objects or alter names extra work will be needed elsewhere.

File: test/controllers/beneficial.owner.type.controller.spec.ts [hunk 12]
```diff
@@ -38,14 +38,12 @@ import {
   APPLICATION_DATA_NO_TRUSTS_MOCK,
```
Comment (JoshFen95) [inline test/controllers/beneficial.owner.type.controller.spec.ts:38]: I feel that changes in this file are out of scope

File: test/utils/trust/beneficial.owner.mapper.spec.ts [hunk 13]
```diff
@@ -2,7 +2,7 @@ import { beforeEach, describe, expect, jest, test } from "@jest/globals";
 import { mapBoIndividualToPage, mapBoOtherToPage } from "../../../src/utils/trust/beneficial.owner.mapper";
```
Comment (JoshFen95) [inline test/utils/trust/beneficial.owner.mapper.spec.ts:2]: Changes in file are out of scope

File: test/utils/trust/historical.beneficial.owner.mapper.spec.ts [hunk 14]
```diff
@@ -29,7 +30,7 @@ describe('Historical Beneficial Owner page Mapper Service', () => {
         const mockFormData = {
           ...mockFormDataBasic,
           boId: '9999',
-          type: '0' as TrustHistoricalBeneficialOwnerType,
+          type: BeneficialOwnerType.Individual as TrustHistoricalBeneficialOwnerType,
```
Comment (JoshFen95) [inline test/utils/trust/historical.beneficial.owner.mapper.spec.ts:33]: I feel like this could also be out of scope

File: src/model/trust.model.ts (hunk 4 above)
Comment (okoye-companieshouse) [reply src/model/trust.model.ts:22]: Following the coding standards, noticed that the keys are added as an array of strings. This did not follow the standard and being a file I was working on I had to correct it. Makes no difference as it is declared but never used can also delete it.

File: test/utils/trust/historical.beneficial.owner.mapper.spec.ts (hunk 14 above)
Comment (okoye-companieshouse) [reply test/utils/trust/historical.beneficial.owner.mapper.spec.ts:33]: I believe this makes the code cleaner as at first glance you do not know what the string '0' means

File: src/model/trust.model.ts (hunk 4 above)
Comment (SteveHicksCH) [reply src/model/trust.model.ts:22]: Please rollback this change.

The keys are used for marshalling data to/from web forms (except for Trust web pages) using these keys and some functions. This also applies for the form that has the Trust spreadsheet  which is processed in `src/controllers/trust.information.controller.ts` but not to the web forms that we are developing in the Trust Screens development.

File: src/utils/trust/individual.trustee.mapper.ts (hunk 11 above)
Comment (SteveHicksCH) [reply src/utils/trust/individual.trustee.mapper.ts:33]: Seconding @JoshFen95 coments on any changes to this model.

Apart from trusteeId (that is only used in the Web application) other fields will need changing in `api-sdk-node`, `overseas-entities-api` and `chips-filing-consumer`. See [Changes to the Historical Beneficial Owners Trustee type](https://companieshouse.atlassian.net/wiki/spaces/TS/pages/4051108053/Changes+to+the+Historical+Beneficial+Owners+BO+Trustee+type) as an example. When changes are requred then we need to get new stories raised to update the other microservices

File: src/model/trust.model.ts [hunk 15]
```diff
@@ -28,20 +29,22 @@ export interface Trust {
...
+  id?: string;
+  type: BeneficialOwnerTypeChoice | RoleWithinTrustType;
   forename: string;
   other_forenames: string;
   surname: string;
   date_of_birth_day: string;
   date_of_birth_month: string;
   date_of_birth_year: string;
   nationality: string;
+  second_nationality?: string;
```
Comment (JoshFen95) [inline src/model/trust.model.ts:47]: This will also need to be added in the overseas-entities-api and chips-filing-consumer. Add a note to the Jira and tag Leah and Sarah so future stories can be added

Comment (SteveHicksCH) [reply src/model/trust.model.ts:47]: Also this will need to be added to the `api-sdk-node` otherwise the `overseas-entities-api` won't get it. As well asking Leah and Sarah for a new Jira I think we need the Test story for this slice to emphasis for the testers that these are new optional fields and they need to do focused testing on it

File: src/model/trust.model.ts [hunk 16]
```diff
@@ -78,6 +78,19 @@ interface TrustHistoricalBeneficialOwnerCommon {
...
+  type: RoleWithinTrustType.BENEFICIARY | RoleWithinTrustType.GRANTOR | RoleWithinTrustType.SETTLOR;
+}
+
+interface InterestedIndividualPersonTrustee extends TrustIndividual{
+  type: RoleWithinTrustType.INTERESTED_PERSON;
+  date_became_interested_person_day: string;
+  date_became_interested_person_month: string;
+  date_became_interested_person_year: string;
+}
+
```
[the remark this answers is not in the document]
Comment (SteveHicksCH) [reply src/model/trust.model.ts:93]: Looks good from point of adding more type information to our code.

File: src/model/trust.page.model.ts [hunk 17]
```diff
@@ -35,13 +36,42 @@ interface TrustHistoricalBeneficialOwnerFormCommon {
...
+  date_of_birth_year: string,
+  nationality: string,
+  second_nationality?: string,
+  property_name?: string,
+  address_line1: string,
+  address_line2: string,
+  city: string,
+  county: string,
+  country: string,
+  postal_code: string,
```
Comment (JoshFen95) [inline src/model/trust.page.model.ts:55]: After looking back at mine I think:
-`property_name?: string,` should not be optional. As it links to premise which is essential the house name or number(wierd naming)
- `county: string,` I think should be optional as marked so on the form. (I updated this myself)
- `postal_code: string,`  I think should be optional as it's marked as "if applicable" on the form. (I updated this myself)
Here is the link to the prototype for you too see - https://roe-v13.herokuapp.com/register/trusts/trust-individual

These changes should be reflected in the `trust.model.ts` too

Happy to approve after this

Comment (okoye-companieshouse) [reply src/model/trust.page.model.ts:55]: thanks missed it

**Review by JoshFen95 — APPROVED** [review]
